### ISSUE TITLE
Error getting lock status

### ISSUE THREAD

**Issue by Tiger519** [nfarina/homebridge-kevo#9]
My locks are frequently showing "Not Responding", and I'm seeing the following in homebridge log:

[7/17/2017, 3:34:02 PM] [Laundry Room Door] Already logged in.
[7/17/2017, 3:34:03 PM] [Laundry Room Door] Error getting lock status: Error: Invalid status code 500
[7/17/2017, 3:34:03 PM] [Laundry Room Door] Lock status is undefined
[7/17/2017, 3:34:03 PM] [Laundry Room Door] Error getting state: Error: Invalid lock status 'undefined'
  Accessory [HomeBridge] Got Characteristic "Lock Current State" value: undefined +459ms
  Accessory [HomeBridge] Error getting value for Characteristic "Lock Current State": Invalid lock status 'undefined' +3

The thing is, when I tap on the lock, it tries again and usually brings it back. I am currently running Node v6.11.1 on a Raspberry Pi Zero W.

Curiously, when I was running on my Windows 10 PC, I almost never encountered this issue over the course of a month or so. However, my server connections were always closing and making my devices not responsive until I reset Homebridge. I don't have that issue on my Pi, but now I seem to be frequently getting these other issues. I was running on Node 5.something before, but I don't imagine that is causing the issue.

EDIT: I have 4 locks, and even though I didn't have this problem when running homebridge on  Windows 10, I'm wondering if it could be related to the rapid requests issue #4, #6.

**Comment by apwelsh** [comment]
I will have to get my Kevo hooked up via linix to see if I can reproduce this. But, I only have 1 lock.  That may be a factor....

**Comment by Tiger519** [comment]
What's perplexing is that the change from Windows to Linux shouldn't have any impact on whether lock status is returned. I did more testing tonight and it never failed when I was querying the locks via Siri, but between 1-4 (usually 3) locks failed to return a status when opening up the Home app or control center. So that definitely seems to be a factor. I just don't understand what about the change from Windows to Linux could cause it.

I'd run on Windows if I could, but I'm pretty sure the Bonjour service keeps crashing (my iTunes sharing goes down too, and sometimes restarting iTunes would make Homebridge come alive again). I was running it as a service with a restart every 2 minutes... last straw was when Windows randomly stopped letting me install it as a service after one of the updates.

Anyway, I appreciate you taking a look at this!

**Comment by apwelsh** [comment]
My bad. Bonjour would not impact this issue.    I misread it originally and followed up that I will try to get a Linux version up and running but not sure if I can reproduce yhe issue since I only have one lock.    My guess is it may be related to the asynchronous nature of Node.js, and that it could be that you are getting more than one concurrent web request to the kevo api.    Maybe I should look at the source first.    HahA
  

  
  

  
  
>

**Comment by Tiger519** [comment]
I was able to decipher the API from the source and got the commands working manually. I can build the URLs and make the calls in my browser, which bring back the lock info JSON as well as remotely lock and unlock the deadbolt.

If I manually try to pass MULTIPLE lock IDs, (i.e. https://www.mykevo.com/user/remote_locks/command/lock.json?arguments=LOCKID1,LOCKID2), I'm getting an HTTP 500 error. But I'm assuming based on the logs this is being attempted one by one. Still, I'm not sure if line 159 in the source code is trying to concatenante as I wrote above, or if the function knows to cycle through them (or if that's a function of HomeKit). Perhaps the issue is that the request is initiating in HomeKit simultaneously, which is somehow causing the token to reset midstream (when I load login page with GET in SoapUI the token changes every time) and cause subsequent actions to fail? If I submit a request without logging in, I also get the HTTP 500 error (which you would somewhat expect), so perhaps something is happening on that end with the multiple requests.

I've updated my "favorites" to include only one lock, and right up until I was about to type this it was working without fail. But now I'm having the same issues with the one lock as well (the others are still connected in the background). So now I'm not even sure the multiple locks are the issue.

I should also note that I have 4 TPLink lightbulbs connected, and those are working fine. I've added them all to my favorites and they simultaneously load without fail every time, so I definitely think that there is some issue specifically with Kevo (either the plugin or the Kevo API itself). If I have time, I will try to get this running on my Windows PC again and see if I'm experiencing the same issues (it's been a few weeks since I had it up on that machine). I can also try updating my config.json to only have one lock and try that on the Pi for a while. But I would appreciate any other insight and testing that you can offer to help track this down!

EDIT: One last note, while I was tinkering I was able to get this to work in SoapUI by capturing the cookie from the login page and passing it in the header of the signin action, and the capturing the cookie from the signin response page and passing it in the header of the info/lock/unlock requests. I'm guessing however this is implemented in the code, that cookie is getting stored somehow, but I can't figure out where. If either of these cookies is missing or incorrect, I get the 500 response. Seems login is working fine but I tend to get the error on accessing lock info.

**Comment by Tiger519** [comment]
Apologies for all the comments, but I've found some additional information that is likely important.

- This is NOT working in Windows 10 as I stated previously. It was ~3 weeks ago, but I started the server up again today and I'm getting the same errors.
- This is NOT related to multiple locks. I updated my JSON to have only one lock (and removed the lightbulbs, so lock was only accessory) and I was still having the issues.
- This is NOT related to NodeJS version. I was running v5.0.0 on Win10 and v6.11.1 on my RPi with the same issues.

The last tidbit I'll leave you with is that when I open the Home app or do a mass query, every action seems to be done twice. I'm actually getting the right statuses back in the overall response, but it seems the fact that ONE of the queries fails makes it show up as not responsive. I have confirmed that when BOTH requests show up as good, I'm getting a status in the Home app.

[7/18/2017, 4:23:21 PM] [Garage Entry Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Bedroom Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Bedroom Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Laundry Room Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Laundry Room Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Front Door] Getting current state...
[7/18/2017, 4:23:21 PM] [Front Door] Getting current state...
[7/18/2017, 4:23:22 PM] [Laundry Room Door] Error requesting login page: Error: Invalid response code 500
[7/18/2017, 4:23:22 PM] [Garage Entry Door] Already logged in.
[7/18/2017, 4:23:22 PM] [Laundry Room Door] Already logged in.
[7/18/2017, 4:23:22 PM] [Front Door] Already logged in.
[7/18/2017, 4:23:22 PM] [Garage Entry Door] Error requesting login page: Error: Invalid response code 500
[7/18/2017, 4:23:22 PM] [Front Door] Error requesting login page: Error: Invalid response code 500
[7/18/2017, 4:23:22 PM] [Bedroom Door] Already logged in.
[7/18/2017, 4:23:22 PM] [Bedroom Door] Already logged in.
[7/18/2017, 4:23:22 PM] [Bedroom Door] Error getting lock status: Error: Invalid status code 500
[7/18/2017, 4:23:22 PM] [Bedroom Door] Lock status is undefined
[7/18/2017, 4:23:22 PM] [Bedroom Door] Error getting state: Error: Invalid lock status 'undefined'
[7/18/2017, 4:23:22 PM] [Laundry Room Door] Lock status is Locked
[7/18/2017, 4:23:22 PM] [Garage Entry Door] Lock status is Unlocked
[7/18/2017, 4:23:22 PM] [Bedroom Door] Lock status is Locked
[7/18/2017, 4:23:22 PM] [Front Door] Lock status is Locked

**Comment by frankea** [comment]
I'm seeing a lot of similar messages:
[7/19/2017, 12:02:37 PM] [Back Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Back Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Front Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Front Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Garage Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Garage Door] Getting current state...
[7/19/2017, 12:02:37 PM] [Front Door] Error requesting login page: Error: Invalid response code 500
[7/19/2017, 12:02:37 PM] [Garage Door] Error requesting login page: Error: Invalid response code 500
[7/19/2017, 12:02:37 PM] [Back Door] Error requesting login page: Error: Invalid response code 500


This is on macOS 10.13 Beta (17A306f) with node 8.1.4.

**Comment by Tiger519** [comment]
@frankea Are you still seeing these errors? Since the evening of 7/20 I have not seen any issues. I did end up rebooting my RPi a few times (tried to switch it to my 5GHz network for better performance without realizing the Pi doesn't support 5Ghz networking! And didn't have it near a screen to be able to fix...), so I don't know if that had anything to do with it, but more likely it was some temporary issue that has been resolved, possibly either with the Kevo website itself or the HTTP request module that enables management of session cookies within the homebridge-kevo module. Not sure if it's better or worse if the issue is not related to a single machine, as really both are quite difficult to troubleshoot.

I should note that in manual testing thorough SoapUI, I did manage to cause some HTTP 500 errors on my "request source" submission, which is nothing more than loading the https://www.mykevo.com/login site (so there should be absolutely no reason for that to fail). But I'm no webservice expert and certainly not an expert on session cookies and tokens, so there could have been other reasons for those failures (and it was probably only 2 times in 100 tries, not on every request like was happening with Homebridge).

I'll leave the issue open for at least a few more days to make sure the problem doesn't reoccur, but everything for the past 4 days has been fine at least for me.

@apwelsh Do you have any thoughts on this and why the issue seemingly has resolved itself?

**Comment by dishguy** [comment]
im getting the same error on pi zero as well with only 1 lock... btw the kevo lock app will give you the lock id...

**Comment by Tiger519** [comment]
@frankea Did you every get your locks working? Mine were rock solid for about 8 months but I started having issues again at the end of March. From all I can tell, this is an issue on Kevo's side, intentional or not, that is causing these requests to fail. If I log in through my browser and hit the .json commands, I intermittently get the same errors. Really frustrated that the setup was working so good. My wife loves the way these locks look, and I don't want to shell out $800 for new locks, so I guess I'm going to stick with them but I feel like I lose so much but not being able to have reliable automations anymore...

**Comment by frankea** [comment]
@Tiger519 To be honest, I just gave up on homebridge altogether.  The main issue really being the Kevo locks.  I was planning on seeing if I can get rid of them and upgrade to actual HomeKit compatible locks, but I've been a bit lazy about it and have just been using them as normal locks. :\

**Comment by Tiger519** [comment]
I'm going to go ahead and close the issue since it's the same issue from #11 and there is a workaround posted in that issue. Still needs a permanent fix if anyone is so inclined. If I can learn a bit more about nodejs and npm (and github) I might tackle an update at some point, for now I've made the changes as discussed in the other issue on my local.

TL;DR - Issue appears to be caused (perhaps intentionally) by Kevo website, temporary fix using requestretry detailed in issue #11, which is still open with the hope that someone can make this a permanent fix for all users.
